### PR TITLE
DNM: rgw dmclock integration

### PR DESCRIPTION
**Pull request by theanalyst**
Just a wip-version of Casey's dmclock branch for discussion and the like, depends on https://github.com/ceph/dmclock/pull/50

### CODE REVIEW

File: src/rgw/rgw_op.h [hunk 1]
```diff
@@ -183,6 +184,10 @@ class RGWOp : public DoutPrefixProvider {
   std::ostream& gen_prefix(std::ostream& out) const override;
   CephContext* get_cct() const override { return s->cct; }
   unsigned get_subsys() const override { return ceph_subsys_rgw; }
+
+  using client_id = rgw::dmclock::client_id;
+  virtual client_id dmclock_client() { return client_id::metadata; }
+  virtual rgw::dmclock::Cost dmclock_cost() { return 0; }
```
Comment (theanalyst) [inline src/rgw/rgw_op.h]: is the dmclock_cost supposed to be implemented by the SubOp classes? A 0-cost would eventually hit the assert in RequestTag calculation as we assert everywhere that cost is +ve

File: src/rgw/rgw_dmclock.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+enum class client_id {
+  admin, //< /admin apis
+  auth, //< swift auth, sts
+  data, //< PutObj, GetObj
+  metadata, //< bucket operations, object metadata
+
+  count
+};
+
+using Cost = double; // may change to uint64_t
```
Comment (cbodley) [inline src/rgw/rgw_dmclock.h]: can now be changed to `using crimson::dmclock::Cost;`

File: src/rgw/rgw_op.h (hunk 1 above)
Comment (cbodley) [reply src/rgw/rgw_op.h]: oh yeah, that assert was added in https://github.com/ceph/dmclock/pull/46, so the `return 0;` here should default to 1 instead

File: src/test/rgw/test_rgw_dmclock_scheduler.cc [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+    EXPECT_EQ(boost::system::errc::success, ec1);
+    EXPECT_EQ(PhaseType::reservation, p1);
+
+    auto p2 = queue.async_request(client_id::auth, {}, get_time(), 1, yield[ec2]);
+    EXPECT_EQ(boost::system::errc::success, ec2);
+    EXPECT_EQ(PhaseType::priority, p2);
+  });
+
+  context.poll();
+  EXPECT_TRUE(context.stopped());
```
Comment (theanalyst) [inline src/test/rgw/test_rgw_dmclock_scheduler.cc]: seeing a failure on this test suite where this condition seems false

**Comment by theanalyst** [comment]
jenkins test this please

**Comment by theanalyst** [comment]
jenkins test this please

**Comment by theanalyst** [comment]
jenkins test this please

File: src/common/options.cc [hunk 4]
```diff
@@ -6345,6 +6345,93 @@ std::vector<Option> get_rgw_options() {
...
+    .set_default(1024)
+    .set_description("Maximum number of concurrent HTTP requests.")
+    .set_long_description(
+        "Maximum number of concurrent HTTP requests that the beast frontend "
+        "will process. Tuning this can help to limit memory usage under heavy "
+        "load.")
+    .add_tag("performance")
+    .add_see_also("rgw_frontends"),
+
+    Option("rgw_enable_dmclock", Option::TYPE_BOOL, Option::LEVEL_ADVANCED)
```
Comment (cbodley) [inline src/common/options.cc]: nit: rgw_dmclock_enabled to match other rgw_dmclock_ config

**Comment by theanalyst** [comment]
#26008 will be tracking this

**Comment by Fibird** [comment]
I want to know why not implement dmclock_client side on rgw or distributing version of dmclock ? Current implementation is only apply mclock(single node version) into rgw.

**Pull request closed without merging**